From fea326834a88001e066bd0476f5a41f42982aecf Mon Sep 17 00:00:00 2001
From: Yixing Liu <liuyixing1@huawei.com>
Date: Thu, 3 Nov 2022 18:49:27 +0800
Subject: [PATCH] RDMA/hns: Fix device ex_cmd capacity
MIME-Version: 1.0
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: fea326834a88001e066bd0476f5a41f42982aecf
Modified-by-SEL: Yes, modfied since b8e3130dd is merged

Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

driver inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I5YZ0Q

----------------------------------------------------------

When the driver does not have an ex_cmd flag, related ioctl()
or syscall() will fail to execute. for example, if the modify_qp
flag is missing, the following error will appear when running
perftest:
Failed to modify QP to INIT, ret=95
Failed to modify QP to INIT
Failed to modify QP 24 to RTR
Failed to modify QP 25 to RTR
Unable to Connect the HCA's through the link
Unable to Connect the HCA's through the link

So add query_device、create_cq、modify_qp and create_qp flag.

Fixes: df0651079380 ("RDMA/hns: Enable modify_cq for uverbs.")
Signed-off-by: Yixing Liu <liuyixing1@huawei.com>
Reviewed-by: Yangyang Li <liyangyang20@huawei.com>
Reviewed-by: Yue Haibing <yuehaibing@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 drivers/infiniband/hw/hns/hns_roce_main.c |    7 +++++++
 1 file changed, 7 insertions(+)

--- a/drivers/infiniband/hw/hns/hns_roce_main.c
+++ b/drivers/infiniband/hw/hns/hns_roce_main.c
@@ -655,6 +655,13 @@ static int hns_roce_register_device(stru
 		(1ULL << IB_USER_VERBS_CMD_CREATE_AH) |
 		(1ULL << IB_USER_VERBS_CMD_DESTROY_AH);
 
+	ib_dev->uverbs_ex_cmd_mask |=
+				(1ULL << IB_USER_VERBS_EX_CMD_QUERY_DEVICE) |
+				(1ULL << IB_USER_VERBS_EX_CMD_MODIFY_CQ) |
+				(1ULL << IB_USER_VERBS_EX_CMD_CREATE_CQ) |
+				(1ULL << IB_USER_VERBS_EX_CMD_MODIFY_QP) |
+				(1ULL << IB_USER_VERBS_EX_CMD_CREATE_QP);
+
 	if (hr_dev->caps.flags & HNS_ROCE_CAP_FLAG_REREG_MR) {
 		ib_dev->uverbs_cmd_mask |= (1ULL << IB_USER_VERBS_CMD_REREG_MR);
 		ib_set_device_ops(ib_dev, &hns_roce_dev_mr_ops);
